### ISSUE TITLE
Incorrect time.sleep

### ISSUE THREAD

**Issue by eoksum** [pyscript/pyscript#324]
**Describe the bug**
Using time.sleep() is not giving the desired result. I tested it with time.sleep(20) and execution was instant.

**To Reproduce**
Try executing the code below in py-script block:
<pre>
import time

print("Now we will wait 20 seconds...")
time.sleep(20)
print("It finished.")
</pre>
*It finished* will be displayed instantly on page.

**Expected behavior**
20 seconds were expected to elapse before "It finished" message gets displayed.

**Comment by JeffersGlass** [comment]
You'll want to use asyncio.sleep(). See, for instance, the [simple_clock example](https://github.com/pyscript/pyscript/blob/cc6d6247a8917c8aefcd60a62e80671fc6657f5e/pyscriptjs/examples/simple_clock.html).

**Comment by antocuni** [comment]
@JeffersGlass is correct but, at the same time, we should have a better story for non-async stuff.
It should either work as expected or fail loudly. Failing silently as it does now is not a good idea.

**Comment by sentairanger** [comment]
I was able to use asyncio with no problem using this example script:

`import asyncio
  
  async def main():
      for i in range(4):
          print("Motor on")
          await asyncio.sleep(1)
          print("Motor off")
          await asyncio.sleep(1)
  
  await main()`

Hopefully time is resolved so that asyncio.sleep is no longer needed for these kind of tasks.

**Comment by marimeireles** [comment]
@antocuni do you know why it's failing like this?

**Comment by philippjfr** [comment]
>@antocuni do you know why it's failing like this?

I believe browser's simply do not provide a synchronous blocking API. You could write a while loop that keeps the process busy but that's horrific in practice because it locks up the browser tab for the duration of the loop and pegs CPU at 100%.

**Comment by rth** [comment]
Also related to https://github.com/pyodide/pyodide/issues/2354

**Comment by pmp-p** [comment]
hi @antocuni 
> but I bet it's a pyodide/emscripten issue

it's more a python vm loop stack preservation problem but there are good news https://github.com/ethanhs/python-wasm/issues/12#issuecomment-986688520

Though i don't think a sync sleep buried in a C call will ever work **in a simple** without changes to wasm vm cntrol flow itself.

**Comment by ramalho** [comment]
Why not this solution?

```python
import asyncio

def sleep(duration):
    async def async_sleep(duration):
        await asyncio.sleep(duration)
    asyncio.run(async_sleep(duration))


```

**Comment by pmp-p** [comment]
> Why not this solution?

because asyncio.run is a blocking function ( except on pygame-wasm )

**Comment by hoodmane** [comment]
In particular, the problem is that there is only one thread, so if `asyncio.run` were allowed to block, it would prevent events on the event loop from ever being processed so `asyncio.sleep`'s future would never resolve. Instead, the Pyodide implementation of `asyncio.run` schedules the task but doesn't block. Perhaps we should raise an error instead.

**Comment by ramalho** [comment]
But isn't the point of `time.sleep()` to block?

**Comment by ramalho** [comment]
Ah, OK. The problem is there's just one thread. Thanks!

**Comment by hoodmane** [comment]
Apparently the newest version of Emscripten has support for stack switching (thanks to @antocuni for noticing, they didn't even make a changelog entry for it). This should give us a way to fix this once we update to newest Emscripten. Though we might have to fix some Emscripten bugs first.
https://github.com/emscripten-core/emscripten/pull/16779

**Comment by hoodmane** [comment]
Reading more carefully, it seems that using stack switching is pretty far off: we need to wait for the WebAssembly stack switching proposal to be finished and adopted by most browsers. So probably at least a couple years.

**Comment by antocuni** [comment]
> it's more a python vm loop stack preservation problem but there are good news [ethanhs/python-wasm#12 (comment)](https://github.com/ethanhs/python-wasm/issues/12#issuecomment-986688520)

I'm not sure this solves the issue. It might solve it in very simple cases in which we always stay in python, but as soon as you have a callback called from C (or Javascript!) you have the same problem. E.g., the following code:
```python
def square(x):
    time.sleep(1)
    return x*x

def calc_squares(lst):
    return list(map(square, lst))

def main():
    lst = [1, 2, 3, 4]
    print(calc_squares(lst))
```

The call from `main` to `calc_square` can be eliminated, but then we call `map()` (which is in C) which calls `square`. So when we call `time.sleep()` the C stack is roughly this:
```
timemodule.c:pysleep()
ceval.c:PyEval_EvalFrame(frame = square)
bltinmodule.c:map_next()
listobject.c:list___init___()
ceval.c:PyEval_EvalFrame(frame = calc_squares)
### NOTE: no EvalFrame for main(), because it was optimized away
```

So, in order to yield the control back to the browser loop, we need to fully switch the whole C stack. The more I think about it, the more I am convinced that we need either asyncify or the proper WASM-level support for stack switching.

Note that `unthrow` has basically the same limitation: it might work as long as you are in Python-only land, but it fails as soon as you have a Python->C->Python callback (which are more common than one might think).

**Comment by hoodmane** [comment]
Agreed, either stack switching or something like synclink or -pthreads is needed. Stack switching has the fewest limitations, for instance it's impossible to link a wasm module using synclink or pthreads but it's possible to do so with stack switching.

So for each one there's a limitation:
* synclink: complex setup
* pthreads: doesn't work with dynamic linking yet
* Stack switching: feature not available yet

**Comment by alexmojaki** [comment]
[sync-message](https://github.com/alexmojaki/sync-message) provides primitives for synchronous communication with web workers, which includes the ability to sleep. But it requires either registering a (simple) service worker or enabling SharedArrayBuffer with cross origin isolation.

https://github.com/alexmojaki/comsync, https://github.com/alexmojaki/pyodide-worker-runner, and https://github.com/alexmojaki/python_runner build on sync-message in layers for greater ease of use and other features. You can see them in action in https://futurecoder.io/course/#ide or https://papyros.dodona.be/ where `time.sleep()` and `input()` work.

**Comment by marimeireles** [comment]
@antocuni the idea is that we tackle this, correct? Or are we're closing it as resolved?

**Comment by antocuni** [comment]
It's not resolved, so it should be kept open.

**Comment by WebReflection** [comment]
Current PyScript offers a `xworker.sync.sleep` utility in workers that works as expected, either *sync* or *async* code, it doesn't matter.

For other cases, the `sleep` must be awaited as previously suggested as we can't block out of the box the whole main thread while in workers that's never an issue.

Closing as addressed or solved already with latest PyScript.
